# Assignment 5: Association mining

## Objective of this assignment
The overall objective is to understand how frequent itemsets can be extracted by
the Apriori algorithm and be able to calculate and interpret association rules in terms of support and confidence.

## ** Important: ** When handing in your homework:
+ Hand in the notebook (and nothing else) named as follows: StudentName1_snumber_StudentName2_snumber.ipynb
+ Provide clear and complete answers to the questions below under a separate header (not hidden somewhere in your source code), and make sure to explain your answers / motivate your choices. Add Markdown cells where necessary.
+ Source code, output graphs, derivations, etc., should be included in the notebook.
+ Hand-in: upload to Blackboard.
+ Include name, student number, assignment (especially in filenames)!
+ When working in pairs only one of you should upload the assignment, and report the name of your partner in your filename.
+ For problems or questions: use the BB discussion board or email the student assistants.


## Advised Reading and Exercise Material
**The following reading material is recommended:**

- Pang-Ning Tan, Michael Steinbach, and Vipin Kumar, *Introduction to Data Mining*, section 6.


## Additional Tools
For this exercise you will need to load the provided *apriorimining.py* script. 


##  5.1 Association mining for course data 
We will use the Apriori algorithm to automatically mine for associations. The Apriori algorithm is an adapted version of the script found here: https://github.com/nalinaksh/Association-Rule-Mining-Python

Check out the script and doc and check if you understand how the association rules are computed. 



In [ ]:
import Toolbox.apriorimining


#### 5.1.1

(0 points) Look at the data file `Data/courses.txt` into Python. The data is represented in Table 1. Inspect the file Data/courses.txt and make sure you understand how the data in Table 1 is stored in the text file.

##### Table 1
|#  |   History |Math| Biology| Spanish | Economics| Physics | Chemistry | English  |  
| :-------------: |:-------------:| :-----------:| :----------:| :------------:|:-------------:| :------------:|  :-------------: | :-------------: |
|student 1 | 0| 1 | 0 | 0 | 1| 1 |1 |1   
|student 2 | 1| 1 | 1 | 0 | 0| 1 |1 |1   
|student 3 | 0| 1 | 0 | 1 | 0| 1 |0 |1   
|student 4 | 0| 0 | 1 | 0 | 0| 1 |1 |0   
|student 5 | 0| 1 | 0 | 0 | 0| 1 |1 |0        
|student 6 | 0| 1 | 1 | 0 | 0| 1 |1 |1   


#### 5.1.2
(1 point) We will analyze the data in Table 1 automatically using the function `apriorimining.generate_association_rules()` from the script. Analyze the data with $ minsupport  \geq 80 \% $ and $ minconfidence \geq 100 \%$.What
are the generated association rules? What kind of conclusions can you make based on these association rules about the subjects that students took?  
  


In [ ]:
import itertools
from Toolbox import apriorimining
with open('Data/courses.txt') as f:
    content = f.readlines()
content = [x.strip() for x in content] 

apriorimining.generate_association_rules()

*Double click to type your answer for 5.1.2 here:*
In Rule #1, regradless on course student take, evry student choose 6(Pysics).
In Rule #2, the percentage of student who choose 2(Math) and 6(Pysics) is 83.33(5/6) and if one student chose 2(Math), he also choose 6(Pysics)
In Rule #3, the percentage of student who choose 7(Chemistry) and 6(Pysics) is 83.33(5/6) and if one student chose 7(Chemistry), he also choose 6(Pysics)


  ##  5.2 Association mining for MovieLens data 
  
  
  In this part of the exercise we consider a Market Basket data set containing 943 users purchases of 1682 movies. A total of 100,000 movies
have been purchased.The data set is called MovieLens100K and is provided by http://www.grouplens.org/node/73, see also the readme `MovieLensData.txt` in the data folder. The data currently considered is not the original data but modified for the apriori algorithm.

#### 5.2.1
  (0 points) The MovieLens data is stored in the file MovieLensData.txt. Inspect the file to see how the data is stored.


#### 5.2.2 
  (1 point) Find association rules using the function below with $ minsupport  \geq 30 \% $ and $ minconfidence \geq 80 \%$. What are the associations with strongest confidence? Do these associations make sense? The script can use file Data/u.item to print the movie titles in stead of numbers. If you enter filename `MovieLensData.txt`, the script will provide an additional option for this. 
  

In [38]:
C1 = {}
transactions = 0
D = []
T = []
with open("Data/MovieLensData.txt", "r") as f:
    for line in f:
        T = []
        transactions += 1
        for word in line.split():
            T.append(word)
            if word not in C1.keys():
                C1[word] = 1
            else:
                count = C1[word]
                C1[word] = count + 1
        D.append(T)
print(D[0])

apriorimining.generate_association_rules()

['1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,57,58,59,60,61,62,63,64,65,66,67,68,69,70,71,72,73,74,75,76,77,78,79,80,81,82,83,84,85,86,87,88,89,90,91,92,93,94,95,96,97,98,99,100,101,102,103,104,105,106,107,108,109,110,111,112,113,114,115,116,117,118,119,120,121,122,123,124,125,126,127,128,129,130,131,132,133,134,135,136,137,138,139,140,141,142,143,144,145,146,147,148,149,150,151,152,153,154,155,156,157,158,159,160,161,162,163,164,165,166,167,168,169,170,171,172,173,174,175,176,177,178,179,180,181,182,183,184,185,186,187,188,189,190,191,192,193,194,195,196,197,198,199,200,201,202,203,204,205,206,207,208,209,210,211,212,213,214,215,216,217,218,219,220,221,222,223,224,225,226,227,228,229,230,231,232,233,234,235,236,237,238,239,240,241,242,243,244,245,246,247,248,249,250,251,252,253,254,255,256,257,258,259,260,261,262,263,264,265,266,267,268,269,270,271,272']
Please enter sup

The rule which has highst confidence(99.66) is #75{ 172, 174, 181 } ==> { 50 }, which occupies 31.18% 
Rule #75: , sup= 31.18, conf= 99.66

#### 5.2.3 
(1 point) Which movies have been watched by the most users? There are only few rules with more than three items. Why?

number 50 is watched most, 60% of people watched it.

#### 5.2.4
(0.5 points) Often we are interested in rules with high confidence. Is it possible for
itemsets to have very low support but still have a very high confidence?

*Double click to type your answer for 5.2.4 here:*


## 5.3 Calculating support, confidence and interest

Calculate these measures and write down how you computed things, not just the answers. 


#### 5.3.1
 Suppose we have market basket data consisting of 100 transactions and 20 items. The support for item $ \text{a} = 45 \%$, the support for item $ \text{b} = 80 \%$ and the support for itemset $ \text{ {a,b }} = 30 \%$. Let the support and confidence thresholds be 20$ \%$ and 60$ \%$, respectively.
  
1. (0.5 points) Compute the confidence of the association rule $ \text{ {a } } \rightarrow   \text{{b }} $. Is the rule interesting according to the confidence measure?

2. (0.5 points) Compute the interest measure (or lift, see slide 44 of chapter 6) for the association pattern $ \text{ {a,b}}$. Describe the nature of the relationship between item $ \text{a}$ and item $ \text{b}$  in terms of the itemset measure.
3. (1 points) What conclusion can you draw from the results of parts (1) and (2)?

*Double click to type your answer for 5.3.1 here:*



#### 5.3.2

1. (1 points) Let $c_1$, $c_2$, and $c_3$ be the confidence values of the rules $ \text{ {a } } \rightarrow   \text{{b }} $, $ \text{ {a } } \rightarrow   \text{{b,c }} $, and $ \text{ {a,c } } \rightarrow   \text{{b }} $ respectively. If we assume that $c_1$, $c_2$, and $c_3$ have different values, what are the possible inequality relationships (e.g. $c_1 \leq c_2 \leq c_3$) among $c_1$, $c_2$, and $c_3$? Which rule has the lowest confidence?
2. (0.5 points) Suppose the confidence of the rules  $ \text{ {a } } \rightarrow   \text{{b }} $ and $ \text{ {b } } \rightarrow   \text{{c }} $ are larger than the confidence threshold. Is it possible that $ \text{ {a } } \rightarrow   \text{{c }} $ has a confidence below that threshold? If no, explain why. If yes, give an example. 

*Double click to type your answer for 5.3.2 here:*




#### 5.3.3

(3 points) Consider the relationships between customers who buy high-definition televisions and exercise machines as shown in Table 2 and 3.

1. Compute the odd ratios for both tables.
2. Compute the $\phi$-coefficient for both tables.
3. Compute the interest (or lift, in the book) factor for both tables.

For Table 3 you should compute measures given above separately for College
Students and for Adults. For each of the measures, describe how the direction
of association changes when data is pooled together (Table 2) instead of being
separated into two groups (Table 3)

##### Table 2: Two way contingency table between the sale of high-definition television and exercise machine
| |   Buy Exercise machine |     |     |
| :------------- | -------------:| :-----------:| :----------:| 
| **Buy HDTV     ** | yes | no | total |
| yes  | 105| 87 | 192 | 
| no | 40| 62 | 102 |   
| total | 145 | 149 | 294 | 
 

##### Table 3: Example of three-way contingency table
| | |   Buy Exercise machine |     |     |
|--- | :------------- | -------------:| :-----------:| :----------:| 
|**Customer group** | **Buy HDTV     ** | yes | no | total |
|College students | yes  | 2| 9 | 11 | 
| | no | 5| 20 | 25 |
| Working adults | yes  | 103| 78 | 181 | 
| | no | 35| 42 | 77 |  



*Double click to type your answer for 5.3.3 here:*
